File: Solution/src/com/solution/dynamicP/MimimumCoinChange.java
package com.solution.dynamicP;

import java.util.Scanner;

import com.solution.main.ReadInput;

public class MimimumCoinChange {

	public void run() {

		Scanner sc = ReadInput.getScanner("coinChange");

		int sum = sc.nextInt();
		int size = sc.nextInt();

		int[] c = ReadInput.getIntArray(size, sc.next());

		int ways = dorun(c, sum);

		System.out.println(ways);
	}

	private int dorun(int[] c, int sum) {

		int[] T = new int[sum + 1];

		for (int i = 1; i < T.length; i++) {
			T[i] = Integer.MAX_VALUE;
		}

		for (int i = 0; i < c.length; i++) {
			for (int j = 1; j < T.length; j++) {
				if(j >= c[i]){
					T[j] = Math.min(T[j],getVal(T[j-c[i]]));
				}
			}
			/*for(int k =0 ; k < T.length ; k++){
				System.out.print(T[k]+" ");
			}
			System.out.println();*/
		}
		
		return T[T.length-1];

	}

	private int getVal(int i) {
		if(i== Integer.MAX_VALUE){
			return Integer.MAX_VALUE;
		}
		return 1+i;
	}
}

File: src/main/kotlin/com/cssolutions/dsalgo/dynamicProgramming/MimimumCoinChange.kt
package com.cssolutions.dsalgo.dynamicProgramming

import com.cssolutions.dsalgo.main.ReadInput

class MimimumCoinChange {
    fun run() {
        val sc = ReadInput.getScanner("coinChange")
        val sum = sc.nextInt()
        val size = sc.nextInt()
        val c = ReadInput.getIntArray(size, sc.next())
        val ways = dorun(c, sum)
        println(ways)
    }

    private fun dorun(c: IntArray, sum: Int): Int {
        val T = IntArray(sum + 1)
        for (i in 1 until T.size) {
            T[i] = Int.MAX_VALUE
        }
        for (i in c.indices) {
            for (j in 1 until T.size) {
                if (j >= c[i]) {
                    T[j] = Math.min(T[j], getVal(T[j - c[i]]))
                }
            }
            /*for(int k =0 ; k < T.length ; k++){
				System.out.print(T[k]+" ");
			}
			System.out.println();*/
        }
        return T[T.size - 1]
    }

    private fun getVal(i: Int): Int {
        return if (i == Int.MAX_VALUE) {
            Int.MAX_VALUE
        } else 1 + i
    }
}